igvc_navigation/src/navigation_server/navigation_server.cpp

```diff
@@ -147,35 +150,82 @@ void NavigationServer::actionGetPathDone(const actionlib::SimpleClientGoalState
         ROS_WARN_STREAM_NAMED("nav_server", "Abort the execution of the planner: " << result.message);
-        current_goal_handle_.setAborted(navigate_waypoint_result, state.getText());
-        current_state_ = FAILED;
+        goal_handle_.setAborted(navigate_waypoint_result, state.getText());
+        navigation_state_ = FAILED;
       }
       break;
     case actionlib::SimpleClientGoalState::PREEMPTED:
       // the get_path action has been preempted.
       // copy result from get_path action
       navigate_waypoint_result.outcome = result.outcome;
       navigate_waypoint_result.message = result.message;
-      current_goal_handle_.setCanceled(navigate_waypoint_result, state.getText());
-      current_state_ = CANCELED;
+      goal_handle_.setCanceled(navigate_waypoint_result, state.getText());
+      navigation_state_ = CANCELED;
       break;
     case actionlib::SimpleClientGoalState::LOST:
       ROS_FATAL_STREAM_NAMED("nav_server", "Connection lost to the action \"get_path\"!");
-      current_goal_handle_.setAborted();
-      current_state_ = FAILED;
+      goal_handle_.setAborted();
+      navigation_state_ = FAILED;
       break;
     default:
       ROS_FATAL_STREAM_NAMED("nav_server", "Unreachable state reached for get_path: " << state.toString());
-      current_state_ = FAILED;
+      navigation_state_ = FAILED;
       break;
   }
+
+  if (navigation_state_ != EXE_PATH)
+  {
+    return;
+  }
+  boost::lock_guard<boost::mutex> guard(replanning_mtx_);
+  replanning_rate_.reset();
+  replanning_rate_.sleep();
+  if (navigation_state_ != EXE_PATH ||
+      action_client_get_path_.getState() == actionlib::SimpleClientGoalState::PENDING ||
+      action_client_get_path_.getState() == actionlib::SimpleClientGoalState::ACTIVE)
+    return;
+  ROS_INFO_STREAM_NAMED("nav_server", "Start replanning, using the \"get_path\" action!");
+  action_client_get_path_.sendGoal(get_path_goal_,
+                                   boost::bind(&NavigationServer::actionGetPathReplanningDone, this, _1, _2));
+}
+
+void NavigationServer::actionGetPathReplanningDone(const actionlib::SimpleClientGoalState &state,
+                                                   const mbf_msgs::GetPathResultConstPtr &result_ptr)
+{
+  if (navigation_state_ != EXE_PATH)
+    return;
+
+  if (state == actionlib::SimpleClientGoalState::SUCCEEDED)
+  {
+    const mbf_msgs::GetPathResult &result = *(result_ptr.get());
+    ROS_DEBUG_STREAM_NAMED("move_base", "Replanning succeeded; sending a goal to \"exe_path\" with the new plan");
+    nav_msgs::Path path = result.path;
+    if (fix_goal_poses_)
+    {
+      size_t size = path.poses.size();
+      // change the orientation of the last pose to that of the pose before it
+      path.poses[size - 1].pose.orientation = path.poses[size - 2].pose.orientation;
+    }
+    runExePath(path);
+  }
+
+  replanning_mtx_.lock();
+  replanning_rate_.sleep();
+  replanning_mtx_.unlock();
```

Comment:
Rather than do this I think you should create a new scope with a guard in it, ie.
```C++
{
  std::lock_guard<std::mutex> guard{replanning_mtx_};
  replanning_rate_.sleep();
}
```